### PR TITLE
[KYUUBI #5306] YarnApplicationOperation supports proxy user

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

For the secured YARN cluster, the Kyuubi Server's user typically has no permission to kill the application. Proxy user or admin should be used instead.

https://docs.cloudera.com/documentation/enterprise/latest/topics/cm_mc_yarn_acl.html#concept_yarn_app_acls__section_killing_an_app

> For YARN, the following three groups of users are allowed to kill a running application:
> - The application owner
> - A cluster administrator defined in yarn.admin.acl
> - A queue administrator defined in aclAdministerApps for the queue in which the application is running

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate

Verified ADMIN mode in internal deployment. (output message is formatted for readable)
```
Error: Batch e351185f-1ed8-437a-91bf-da2174e611e2 failed:
{
    "id":"e351185f-1ed8-437a-91bf-da2174e611e2",
    "user":"da_music",
    "batchType":"SPARK",
    "name":"SparkPi",
    "appStartTime":0,
    "appId":"application_1694730881181_58306",
    "appUrl":"http://xxxx-rm-2.xxxx:8088/cluster/app/application_1694730881181_58306",
    "appState":"KILLED",
    "appDiagnostic":"Application application_1694730881181_58306 was killed by user yarn at 10.49.59.149",
    "kyuubiInstance":"kyuubi-1.kyuubi-headless.spark.svc.cluster.local:10099",
    "state":"CANCELED",
    "createTime":1695102138188,
    "endTime":1695102163341,
    "batchInfo":{}
}
```

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No.

### CODE REVIEW

**Comment by pan3793** [comment]
Also cc @zhouyifan279

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -2826,6 +2826,32 @@ object KyuubiConf {
       .version("1.7.2")
       .fallbackConf(ENGINE_SUBMIT_TIMEOUT)
 
+  object YARNClientProxyUserStrategy extends Enumeration {
+    type YARNClientProxyUser = Value
+    val NONE, ADMIN, OWNER = Value
+  }
+
+  val YARN_PROXY_USER_STRATEGY: ConfigEntry[String] =
+    buildConf("kyuubi.yarn.proxyUser.strategy")
```
Comment (zhouyifan279) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2835]: How about `kyuubi.yarn.client.proxyUser.xxx` ?

**Comment by zhouyifan279** [comment]
LGTM

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 1 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2835]: I'm open to the configuration name, but there are existing `kyuubi.kubernetes.xxx` without `client` namespace too.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala [hunk 2]
```diff
@@ -21,118 +21,144 @@ import java.util.Locale
...
+        }
+      case OWNER =>
+        info("Skip initializing admin YARN client")
+    }
   }
 
-  override def isSupported(appMgrInfo: ApplicationManagerInfo): Boolean = {
-    yarnClient != null && appMgrInfo.resourceManager.exists(
-      _.toLowerCase(Locale.ROOT).startsWith("yarn"))
+  private def createYarnClient(_yarnConf: Configuration): YarnClient = {
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala:64]: `_yarnConf` not used.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala [hunk 3]
```diff
@@ -21,118 +21,144 @@ import java.util.Locale
...
+    yarnConf = KyuubiHadoopUtils.newYarnConfiguration(conf)
+
+    YARNClientProxyUserStrategy.withName(conf.get(KyuubiConf.YARN_PROXY_USER_STRATEGY)) match {
+      case NONE =>
+        val c = createYarnClient(yarnConf)
+        info(s"Successfully initialized admin YARN client.")
+        adminYarnClient = Some(c)
+      case ADMIN =>
+        val adminUser = conf.get(KyuubiConf.YARN_PROXY_USER_ADMIN_USER).getOrElse(Utils.currentUser)
+        Utils.doAs(adminUser) { () =>
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala:54]: We don't need to create a proxy user if the users are the same.

**Review by cxzl25 — APPROVED** [review]
Overall LGTM

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala (hunk 3 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala]: updated

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala (hunk 2 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/YarnApplicationOperation.scala:74]: fixed

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
The last commit renames the configurations

- `kyuubi.yarn.proxyUser.strategy`  => `kyuubi.yarn.user.strategy`
- `kyuubi.yarn.proxyUser.adminUser` => `kyuubi.yarn.user.admmin`

And changes the default value of `kyuubi.yarn.user.admmin` from `None` to `"yarn"`

**Review by wForget — APPROVED** [review]
LGTM

**Comment by pan3793** [comment]
Thanks, merged to master/1.8

**Pull request closed without merging**